### PR TITLE
Add digest verification (sha1) for files in the archive_files.json

### PR DESCRIPTION
**Pull request by smagnuso**
This is an attempt to start verifying files in the archive from network.

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js [hunk 1]
```diff
@@ -484,21 +482,37 @@ var GameArchiveLoader = {
...
+                                    const digest = await window.crypto.subtle.digest("SHA-1", mmap);
+                                    matches = Array.from(new Uint8Array(digest)).map(b => b.toString(16).padStart(2, '0')).join('') == file.sha1;
+                                }
+                            } catch(e) { }
+                            FS.close(stream);
+                        } else {
+                            matches = false;
+                        }
+                    }
+                    if (matches) {
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js:509]: What kind of impact will this verification have on load times? We do not wish to slow down loading of content if we can avoid it. Can we hide this behind an option in game.project?

**Review by britzl — COMMENTED** [review]
See question about impact on load time

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js (hunk 1 above)
Comment (smagnuso) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js:509]: Certainly adds some cost, I've pushed change to have it be optional

Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js:509]: Great, but how do I disable it? We need a game.project setting and a check in HTML5Bundler too don't we?

You need an entry in com/dynamo/bob/meta.properties and in editor/resources/edn.properties. The default value should be off.

**Review by britzl — CHANGES_REQUESTED** [review]
Question about how to disable sha1 verification

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js (hunk 1 above)
Comment (smagnuso) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js:509]: Yes, I made the default value off. I am just passing --with-sha1 into bob.jar currently. Do you want it to show up in the editor?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java [hunk 2]
```diff
@@ -581,6 +581,8 @@ public static List<CommandLineOption> getCommandLineOptions() {
 
                 opt("l", "liveupdate", ONE, "Yes if liveupdate content should be published", true),
 
+                opt(null, "with-sha1", ZERO, "Generate (and verify) sha1 signatures from build artifacts", false),
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java:584]: How will a user know this only applies to html5 builds?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js (hunk 1 above)
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js:509]: Now that I think about, it's probably fine that it is an option only available in command line builds. But the name of the option doesn't tell a user that it only applies to html5 builds. We should at least write that when printing the help.

**Review by britzl — COMMENTED** [review]
See comment about platform availability

**Comment by AGulev** [comment]
This option should be specified in game.rpoejct, not as bob argument 
How it's added (ignore cpp) https://github.com/defold/defold/commit/970042be550ad43b7d71c16c9f290d50b15dd518
and then it will be replaced in dmloader using templater: https://github.com/defold/defold/blob/1645c043da32d8b31397627532897ea974c02935/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/bundle/resources/jsweb/dmloader.js#L40

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java (hunk 2 above)
Comment (smagnuso) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Bob.java]: Added note to description

**Pull request merged**